demo/bulk_insert/bulk_insert.py

```diff
@@ -1,281 +1,339 @@
+        return expected_col_count
+
+    def process_entities(self, expected_col_count):
         for row in self.reader:
-            # Each row should have two columns (a source and dest ID)
-            if len(row) != 2:
-                raise CSVError ("%s:%d Expected 2 columns, encountered %d ('%s')"
-                                % (self.csvfile, self.reader.line_num, len(row), ','.join(row)))
-            for elem in row:
-                # Raise an exception if an element cannot be read as an integer
-                try:
-                    int(elem)
-                except:
-                    raise CSVError ("%s:%d Token '%s' was not a node ID)"
-                            % (self.csvfile, self.reader.line_num, elem))
-        self.total_entities = self.reader.line_num
-
-    def token_count(self):
-        # Relations have 2 tokens: name and pending insertion count.
-        return 2
-
-    def unroll(self):
-        return [self.name, self.pending_inserts]
-
-    def print_insertion_start(self):
-        print('Inserting relation "%s" - %d edges' % (self.name, self.total_entities))
-
-# Issue single Redis query to allocate space for graph
-def allocate_graph(node_count, relation_count):
-    cmd = ["GRAPH.BULK", graphname, "BEGIN", node_count, relation_count]
-    result = redis_client.execute_command(*cmd)
-    print(result)
-
-def finalize_graph():
-    cmd = ["GRAPH.BULK", graphname, "END"]
-    result = redis_client.execute_command(*cmd)
-    print(result)
-
-def query_redis(metadata, entities):
-    cmd = ["GRAPH.BULK", graphname] + metadata.unroll() + entities
-    # Raise error if query doesn't contain entities
-    if not entities:
-        raise Exception ("Attempted to insert 0 tokens( '%s')." % (" ".join(str(e) for e in cmd)))
-
-    # Send query to Redis client
-    result = redis_client.execute_command(*cmd)
-    stats = result.split(', '.encode())
-    metadata.entities_created += int(stats[0].split(' '.encode())[0])
-    metadata.entities_created += int(stats[1].split(' '.encode())[0])
-    metadata.queries_processed += 1
-
-def build_descriptors(csvs, argtype):
-    # Prepare container for all labels
-    arg = Argument(argtype)
-
-    # Generate a label descriptor from each given label CSV
-    for f in csvs:
-        # Better method for this?
-        if (argtype == "NODES"):
-            descriptor = LabelDescriptor(f)
-        else:
-            descriptor = RelationDescriptor(f)
-        arg.descriptors.append(descriptor)
-    arg.total_entities = sum(desc.total_entities for desc in arg.descriptors)
-    return arg
-
-def help():
-    pass
+            self.validate_row(expected_col_count, row)
+
+            try:
+                src = NODE_DICT[row[0]]
+                dest = NODE_DICT[row[1]]
+            except KeyError as e:
+                print("Error - relationship specified a non-existent identifier.")
+                raise e
+            src = NODE_DICT[row[0]]
```

Comment:
define src and dest above try, such that we don't try to access NODE_DICT twice

Reply:
Ah, these two lines can actually just be deleted - `try` doesn't have its own scope in Python.